lib/Conversion/TorchToTcp/Elementwise.cpp

```diff
@@ -107,6 +127,126 @@ class ConvertAtenTanhOp : public OpConversionPattern<AtenTanhOp> {
+      double minFloat;
+      int64_t minInt;
+      if (matchPattern(minValue, m_TorchConstantFloat(&minFloat))) {
+        if (elementType.isa<mlir::FloatType>())
+          minFloatAttr = rewriter.getF32FloatAttr(minFloat);
+        else if (elementType.isa<mlir::IntegerType>())
+          minIntAttr =
+              rewriter.getI64IntegerAttr(static_cast<int64_t>(minFloat));
+      } else if (matchPattern(minValue, m_TorchConstantInt(&minInt))) {
+        if (elementType.isa<mlir::FloatType>())
+          minFloatAttr = rewriter.getF32FloatAttr(static_cast<float>(minInt));
+        else if (elementType.isa<mlir::IntegerType>())
+          minIntAttr = rewriter.getI64IntegerAttr(minInt);
+      } else {
+        return rewriter.notifyMatchFailure(
+            op, "expected to find a float or integer constant as min");
+      }
+    }
+    if (checkNotNone(rewriter, op, maxValue).succeeded()) {
+      double maxFloat;
+      int64_t maxInt;
+      if (matchPattern(maxValue, m_TorchConstantFloat(&maxFloat))) {
+        if (elementType.isa<mlir::FloatType>())
+          maxFloatAttr = rewriter.getF32FloatAttr(maxFloat);
+        else if (elementType.isa<mlir::IntegerType>())
+          maxIntAttr =
+              rewriter.getI64IntegerAttr(static_cast<int64_t>(maxFloat));
+      } else if (matchPattern(maxValue, m_TorchConstantInt(&maxInt))) {
+        if (elementType.isa<mlir::FloatType>())
+          maxFloatAttr = rewriter.getF32FloatAttr(static_cast<float>(maxInt));
+        else if (elementType.isa<mlir::IntegerType>())
+          maxIntAttr = rewriter.getI64IntegerAttr(maxInt);
+      } else {
+        return rewriter.notifyMatchFailure(
+            op, "expected to find a float or integer constant as max");
+      }
+    }
+
+    rewriter.replaceOpWithNewOp<tcp::ClampOp>(op, inputType, input,
+                                              minFloatAttr, maxFloatAttr,
+                                              minIntAttr, maxIntAttr);
+    return success();
+  }
+};
+
+template <typename AtenOpT>
+class ConvertAtenReluOp : public OpConversionPattern<AtenOpT> {
+public:
+  using OpConversionPattern<AtenOpT>::OpConversionPattern;
+  using OpAdaptor = typename AtenOpT::Adaptor;
+
+  LogicalResult
+  matchAndRewrite(AtenOpT op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    Value input = adaptor.getSelf();
+    RankedTensorType inputType = input.getType().dyn_cast<RankedTensorType>();
+    auto elementType = inputType.getElementType();
+    if (!inputType)
+      return rewriter.notifyMatchFailure(
+          op, "Only Ranked Tensor types are supported in TCP");
+    if (!inputType.getElementType().isIntOrFloat())
+      return rewriter.notifyMatchFailure(
+          op,
+          "Clamp input tensor must have integer or floating-point datatype");
+
+    FloatAttr minFloatAttr, maxFloatAttr;
+    IntegerAttr minIntAttr, maxIntAttr;
+
+    if (isa<AtenReluOp>(op)) {
```

Comment:
I'm not a big fan of doing this inside of a template because templates and `isa`s/`dyn_cast`s are each designed to be used under very different circumstances. Templates are for when your code looks exactly the same for all the types involved (with a few exceptions now that C++17 allows things like `if constexpr`) *and* you know all the types involved at compile time. On the other hand, `isa`s/`dyn_cast`s are for when your code does not look the same for all types *and* you don't know all the types involved at compile time.

For this particular case, you do know all the types involved at compile time, so there should be no need for `isa`s in the code. One way to fix this is doing `if constexpr (std::is_same_v<AtenOpT, AtenReluOp>()) ...`. Another way would be to not use templates and just have two classes, one for each `Relu` variant, and extract the code in common (the error checking at the beginning) into a helper function that both `Relu` classes and the `Clamp` class can use.

Reply:
Thanks for those points. I replaced this with different classes for `relu` and `relu6`.

Reply:
Reg extracting the error checking code into a helper function, I'm not sure how that is going to help. I will still have to catch that error message and return it in this function conditionally. Is there a more compact way to do that with error messages in mlir?

Reply:
I was thinking of having something like

```c++
static LogicalResult isIntOrFloatRankedTensor(Operation *op, Value input, PatternRewriter &rewriter);

// In matchAndRewrite
if (failed(isIntOrFloatRankedTensor(op, input, rewriter))
  return rewriter.notifyMatchFailure(...);
```

However, this is not really necessary right now. This can be done later if this check becomes a very common one

Reply:
Okay, sounds good. 

I wish there was something like `RETURN_IF_ERROR`.